ingest/atac.py

```diff
@@ -0,0 +1,43 @@
+"""Process ATAC-seq files for visualization in igv.js
+
+PREREQUISITE:
+- Install htslib: `brew install htslib` on macOS (TODO: Create Docker file)
+
+"""
+import os
+import shutil
+import subprocess
+
+def index_fragments(tsv_path):
+    """Bgzip and TBI-index a raw ATAC-seq fragments file
+    """
+    print("Index fragments file from DSP Pipelines team")
+
+    print("Change suffix from .tsv to .bed, as expected by igv.js")
+    tmp_tsv_path = f"{tsv_path}.tmp"
+    shutil.copy2(tsv_path, tmp_tsv_path) # Preserve original file
+    bed_path = tmp_tsv_path.replace(".tsv.tmp", ".bed")
+    os.rename(tmp_tsv_path, bed_path)
+
+    print("Sort BED file by genomic position")
+    # The "-k1.4V" argument ensures `sort` uses column "1" and breaks ties
+    # by ordering character "4" using version-sort, i.e. "V", which does a
+    # natural sort of (version) numbers within text.  This ensures e.g. "12"
+    # appears after "2".
+    sorted_bed_path = bed_path.replace(".bed", ".possorted.bed")
+    sort_command = (f"sort -k1.4V {bed_path}").split(" ")
```

Comment:
Is there any advantage to using something like [bedtools](https://daler.github.io/pybedtools/autodocs/pybedtools.bedtool.BedTool.sort.html) for this?

Reply:
Thanks for the suggestion!  Yes, `bedtools` has a speed advantage.  And even though it has complexity disadvantage, runtime speed seems more useful than simplicity in this case.  So I've gone ahead and used `bedtools` here (4675c2c).

More specifically, compared to `sort -k1.4V`, `bedtools sort` is:

  * **Faster to run**, by > 60% (14 s) in my local test -- details below.
  * **Harder to debug**.  `bedtools sort` uses lexicographic order, so e.g. chr12 comes before chr2, which is less human-friendly and might confuse someone looking directly in the sorted BED file compared to the more natural order given by `sort -k1.4V`.  But the output is still valid input for `bgzip`, which only requires features to be listed contiguously by chromosome (or scaffold).
  * **An extra dependency to manage**.  Updates, deprecations, and installation are likely to be more expensive with `bedtools`, but probably not by much.

**Speed comparison**
For the 313 MB ATAC-seq fragments file from DSP Pipelines:

```
# bedtools sort
wmce2-f16:atac eweitz$ time bedtools sort -i Library-8-20230710_atac.fragments.bed > Library-8-20230710_atac.fragments.bedtools_sorted.bed

real	0m8.197s
user	0m6.893s
sys	0m0.983s

# DSP Pipelines sort
wmce2-f16:atac eweitz$ time sort -k1,1V -k2,2n Library-8-20230710_atac.fragments.bed > Library-8-20230710_atac.fragments.sort_k11V_k22n.bed

real	0m10.474s
user	0m8.814s
sys	0m1.171s

# Original sort in this PR
wmce2-f16:atac eweitz$ time sort -k1.4V Library-8-20230710_atac.fragments.bed > Library-8-20230710_atac.fragments.sort_k14V.bed

real	0m22.207s
user	0m20.485s
sys	0m1.284s
```

The "DSP Pipelines sort" is from [warp/tasks/skylab/H5adUtils.wdl](https://github.com/broadinstitute/warp/blob/47fb3c33e1dcc9c00e65cd9084ac29857794c2ea/tasks/skylab/H5adUtils.wdl#L271C5-L271C23).  Per data above, `bedtools sort` is > 20% faster than `sort -k1,1V -k2,2n`.  Output seems similar in debuggability; the only disadvantage I see relative to the Pipelines approach is that bedtools would be an extra dependency.  (This might interest you, @ekiernan.)